src/ansi-c/c_typecheck_expr.cpp

```diff
@@ -3594,27 +3725,35 @@ exprt c_typecheck_baset::do_special_functions(
     // do not typecheck the argument - it is never evaluated, and thus side
     // effects must not show up either
 
-    // try to produce constant
-    exprt tmp1=expr.arguments().front();
-    simplify(tmp1, *this);
-
-    bool is_constant=false;
-
-    // Need to do some special treatment for string literals,
-    // which are (void *)&("lit"[0])
-    if(
-      tmp1.id() == ID_typecast &&
-      to_typecast_expr(tmp1).op().id() == ID_address_of &&
-      to_address_of_expr(to_typecast_expr(tmp1).op()).object().id() ==
-        ID_index &&
-      to_index_expr(to_address_of_expr(to_typecast_expr(tmp1).op()).object())
-          .array()
-          .id() == ID_string_constant)
+    bool is_constant = false;
+    if(config.ansi_c.mode == configt::ansi_ct::flavourt::CLANG)
     {
-      is_constant=true;
+      is_constant = is_compile_time_constantt(*this)(expr.arguments().front());
     }
     else
-      is_constant=tmp1.is_constant();
+    {
+      // try to produce constant
```

Comment:
`is_compile_time_constantt` is too far off in the GCC/MSVC case?

Reply:
`__builtin_constant_p` doesn't exist with MSVC, so it'd be GCC only. For that, however, the simplifier appears to be much closer: GCC happily takes `x - x`, `x * 0` as constants. But then we could still do what I now (in my latest push to this PR) ended up doing: produce a variant of `is_compile_time_constantt` for a specific compiler (`clang_is_constant_foldedt`).